### PR TITLE
Add gossip fraud proof capability over p2p

### PR DESCRIPTION
**Pull request by Manav-Aggarwal**
See #439

Since we are adding block fraud proofs, currently the fraud-proof consists of a block height and state witnesses that enable reconstruction of an IAVL tree within cosmos-sdk. The fraud proof struct will be updated after https://github.com/celestiaorg/cosmos-sdk/issues/195 is done.

### CODE REVIEW

File: types/serialization.go [hunk 1]
```diff
@@ -40,6 +40,73 @@ func (h *Header) UnmarshalBinary(data []byte) error {
 	return err
 }
 
+// MarshalBinary encodes Header into binary form and returns it.
+func (fp *FraudProof) MarshalBinary() ([]byte, error) {
```
Comment (tzdybal) [inline types/serialization.go:44]: Could you also add "round trip" (object -marshal-> bytes -unmarshal-> object)  test in `serialization_test.go` to ensure that serialization works as expected?

File: proto/pb/dalc/dalc.pb.go [hunk 2]
```diff
@@ -0,0 +1,1723 @@
+// Code generated by protoc-gen-gogo. DO NOT EDIT.
```
Comment (tzdybal) [inline proto/pb/dalc/dalc.pb.go:1]: This file is moved by the script to other location. Please remove it from PR.

File: proto/pb/optimint/optimint.pb.go [hunk 3]
```diff
@@ -0,0 +1,2407 @@
+// Code generated by protoc-gen-gogo. DO NOT EDIT.
```
Comment (tzdybal) [inline proto/pb/optimint/optimint.pb.go:1]: This file is moved by the script to other location. Please remove it from PR.

File: proto/pb/optimint/state.pb.go [hunk 4]
```diff
@@ -0,0 +1,1027 @@
+// Code generated by protoc-gen-gogo. DO NOT EDIT.
```
Comment (tzdybal) [inline proto/pb/optimint/state.pb.go:1]: This file is moved by the script to other location. Please remove it from PR.

**Review by tzdybal — COMMENTED** [review]
This looks good :+1: 

I have some questions/comments:
 - it seems that files added in `proto/tendermint` are not used - am I right? Could you please remove it?
 - please remove `proto/pb` - it's just a temporary directory (just created: https://github.com/celestiaorg/optimint/issues/453)

**Review by adlerjohn — COMMENTED** [review]
No comments

File: p2p/client_test.go [hunk 5]
```diff
@@ -125,6 +125,58 @@ func TestGossiping(t *testing.T) {
 	wg.Wait()
 }
 
+func TestFraudProofGossip(t *testing.T) {
```
Comment (jbowen93) [inline p2p/client_test.go:128]: I'm not sure it's useful to have this function that's just a copy of `TestGossiping` with `clients[4].GossipTx()` replaced with `clients[4].GossipFraudProof()`

File: proto/optimint/optimint.proto [hunk 6]
```diff
@@ -74,3 +74,13 @@ message Block {
 	Data data = 2;
 	Commit last_commit = 3;
 }
+
+message FraudProof {
+  uint64 block_height = 1;
+  repeated StateWitness state_witnesses = 2;
```
Comment (jbowen93) [inline proto/optimint/optimint.proto:80]: I think it matches the research terminology better to have the `FraudProof` contain a single `StateWitness` that is itself an array of `key/value` pairs. Please correct me if I'm mistaken here.

File: types/block.go [hunk 7]
```diff
@@ -89,3 +89,13 @@ type Signature []byte
 type IntermediateStateRoots struct {
 	RawRootsList [][]byte
 }
+
+type FraudProof struct {
```
Comment (jbowen93) [inline types/block.go:93]: See comment in proto/optimint/optimint.proto

File: types/serialization.go [hunk 8]
```diff
@@ -40,6 +40,73 @@ func (h *Header) UnmarshalBinary(data []byte) error {
...
+// FromProto fills Block with data from its protobuf representation.
+func (fp *FraudProof) FromProto(other *pb.FraudProof) error {
+	var err error
+	fp.BlockHeight = other.BlockHeight
+	fp.StateWitnesses, err = stateWitnessesFromProto(other.StateWitnesses)
+
+	return err
+}
+
+func stateWitnessesToProto(sws []*StateWitness) []*pb.StateWitness {
```
Comment (jbowen93) [inline types/serialization.go:76]: See comment in proto/optimint/optimint.proto

**Review by jbowen93 — COMMENTED** [review]
A few changes in naming/organization but overall LGTM

File: proto/optimint/optimint.proto (hunk 6 above)
Comment (Manav-Aggarwal) [reply proto/optimint/optimint.proto:80]: Added a `StateWitnessData` to consolidate the array of key/value pairs in a struct to reflect your comment

File: types/serialization.go (hunk 1 above)
Comment (Manav-Aggarwal) [reply types/serialization.go:44]: Added. Notice that removing the files in `proto/pb` and `proto/tendermint` lead to linters and build tests failing now.

File: proto/optimint/optimint.proto (hunk 6 above)
Comment (jbowen93) [reply proto/optimint/optimint.proto:80]: I'd prefer to be able to retrieve a single `stateWitness` object that itself containers a slice of `witnessData`
See https://www.notion.so/celestiaorg/Fraud-Proof-Paper-Section-4-07714c3183f944ab8b98394fd3d61287#aaa8337929cb4fc5a8703bc67cb8a1e5

Something like:
```
message FraudProof {
  uint64 block_height = 1;
  StateWitness state_witness = 2;
}

message StateWitness {
   repeated WitnessData witness_data = 1;
}

message WitnessData {
  bytes key = 1;
  bytes value = 2;
}
```

Comment (Manav-Aggarwal) [reply proto/optimint/optimint.proto]: Made changes to reflect this suggestion

File: types/serialization.go [hunk 9]
```diff
@@ -40,6 +40,85 @@ func (h *Header) UnmarshalBinary(data []byte) error {
...
+	return nil
+}
+
+func witnessesFromProto(sw *pb.StateWitness) ([]WitnessData, error) {
+	var witnesses []WitnessData
+	for _, wd := range sw.WitnessData {
+		var witnessData WitnessData
+		err := witnessData.FromProto(wd)
+		if err != nil {
+			return []WitnessData{}, nil
```
[the remark this answers is not in the document]
Comment (Manav-Aggarwal) [reply types/serialization.go]: thanks for the catch!

**Review by tzdybal — APPROVED** [review]
:+1:

**Pull request merged**